### PR TITLE
[V1][Spec Decode] Apply torch.compile & cudagraph to EAGLE

### PR DESCRIPTION
**Pull request by luyuzhe111**
Task 8 of https://github.com/vllm-project/vllm/issues/15901

A few notes regarding the implementation.

### Torch.compile 
1. @support_torch_compile is convenient but requires a specific signature for the model. Changes in `vllm/model_executor/models/llama_eagle.py` address this requirement. 
2. Further, to make torch.compile work, we need a separate cache dir for EAGLE model. Without edits in `vllm/compilation/backends.py`, we wouldn't be able to cache EAGLE's compilation properly since the EAGLE module was registered under vllm config from the target model (see [here](https://github.com/vllm-project/vllm/blob/main/vllm/v1/spec_decode/eagle.py#L226))
3. One notable bug related to torch.compile is this [line](https://github.com/vllm-project/vllm/pull/17211/files#diff-a4809a837fbf535a8f0999b11087a53ec1c53948b50c0a1fe64396bc86de9461R153). Essentially, the default data type for input ids is int32 and the EAGLE model was also compiled with this data type. However, `tensor.argmax()` returns int64 by default. Feeding int64 input ids to the compiled model will completely mess things up and lead to gibberish draft tokens. Currently the compiled model does not even give warnings when the input data type mismatches. Wonder if we can prevent similar bugs in the future by some more checks. 

### CudaGraph
Changes in `vllm/v1/spec_decode/eagle.py` and `vllm/v1/worker/gpu_model_runner.py` are mostly for CudaGraph. Nothing fancy other than registering additional persistent buffers and making sure to use them for EAGLE's forward pass. I do want to mention that with torch.compile & CudaGraph, the EAGLE model's forward pass has been drastically improved (2.5x faster), which makes the small but abundant torch operations look inefficient. Any advice to further optimize these overheads is greatly appreciated.

Finally, it would be great to have https://github.com/vllm-project/vllm/pull/17010 reviewed and merged so that we don't have to pull in other PRs to test the acceptance length. 

Note that the current PR does not directly make torch.compile & cuda graph available for EAGLE3. I think it's worth a separate PR since the work is non-trivial due to the fact that EAGLE-3's input hidden states have dynamic shapes. Maybe @benchislett could help. 

@WoosukKwon @LiuXiaoxuanPKU

### CODE REVIEW

**Comment by WoosukKwon** [comment]
Thanks for the PR. This is so cool!
I’ll take a look, but it would be great if we could also get @youkaichao’s review.

**Review by WoosukKwon — COMMENTED** [review]
@luyuzhe111 Thanks for the PR!

Left some comments. Please take a look!

File: vllm/model_executor/models/llama_eagle.py [hunk 1]
```diff
@@ -37,17 +38,19 @@ def __init__(
...
 
+@support_torch_compile
 class LlamaModel(nn.Module):
 
     def __init__(
         self,
         *,
-        model_config: ModelConfig,
-        start_layer_id: int = 0,
+        vllm_config: VllmConfig,
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/model_executor/models/llama_eagle.py:47]: Oh I didn't know that. thanks for the explanation!

File: vllm/v1/spec_decode/eagle.py [hunk 2]
```diff
@@ -27,10 +27,41 @@ def __init__(
...
         self.block_size = vllm_config.cache_config.block_size
+
+        self.dtype = vllm_config.model_config.dtype
+
+        self.max_num_tokens = vllm_config.scheduler_config \
+            .max_num_batched_tokens
+
+        self.hidden_size = vllm_config.model_config.get_hidden_size()
+
+        # TODO: make eagle3 compatible with cudagraph
```
[the remark this answers is not in the document]
Comment (luyuzhe111) [reply vllm/v1/spec_decode/eagle.py:43]: since the hidden dimension of `hidden_states` as the input to EAGLE-3  can vary ( `hidden_size * 3` or `hidden_size`), there is some extra work that I think deserves a separate PR.

File: vllm/v1/worker/gpu_model_runner.py [hunk 3]
```diff
@@ -1245,18 +1242,28 @@ def execute_model(
...
+                if self.use_cuda_graph and \
+                    num_actual_draft_tokens <= self.cudagraph_batch_sizes[-1]:
+                    num_padded_draft_tokens = self.vllm_config. \
+                        pad_for_cudagraph(num_actual_draft_tokens)
+
+                    if num_padded_draft_tokens > num_actual_draft_tokens:
+                        token_indices = torch.cat((
+                            token_indices,
+                            token_indices[-1].repeat(num_padded_draft_tokens -
+                                                     num_actual_draft_tokens)))
```
[the remark this answers is not in the document]
Comment (luyuzhe111) [reply vllm/v1/worker/gpu_model_runner.py:1256]: ahh really good point, let me double check.

File: vllm/v1/spec_decode/eagle.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py:43]: What do you mean by "vary"? Isn't it fixed after the model is initialized?

[the remark this answers is not in the document]
Comment (luyuzhe111) [reply vllm/v1/spec_decode/eagle.py:43]: see this [line](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama_eagle3.py#L122). apparently the input hidden states can be either from the target model (concatenation of hidden states from three layers) of the EAGLE model (hidden states from a single layer).

File: vllm/v1/worker/gpu_model_runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (luyuzhe111) [reply vllm/v1/worker/gpu_model_runner.py]: @WoosukKwon ah I was doing this since I wanted to make sure that the inputs to `self.drater.propose()` are always padded, which is not optimal actually. please take a look at the updated code. thanks!

File: vllm/model_executor/models/llama_eagle.py [hunk 4]
```diff
@@ -37,17 +38,19 @@ def __init__(
...
         prefix: str = "",
+        start_layer_id: int = 0,
```
[the remark this answers is not in the document]
Comment (luyuzhe111) [reply vllm/model_executor/models/llama_eagle.py:49]: this is just to align with the default non-EAGLE model signature (`vllm_config`, `prefix`, `**kwargs`). :)

File: vllm/v1/worker/gpu_model_runner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (luyuzhe111) [reply vllm/v1/worker/gpu_model_runner.py]: now the inputs are always without padding and the cuda graph padding size will be figured out in `self.drafter.propose()`

File: vllm/v1/worker/gpu_model_runner.py [hunk 5]
```diff
@@ -1500,6 +1493,18 @@ def _dummy_run(
...
+                self.speculative_config.method == 'eagle':
+                assert isinstance(self.drafter, EagleProposer)
+                with set_forward_context(None,
+                                         self.drafter.vllm_config,
+                                         num_tokens=num_tokens):
+                    self.drafter.model(
+                        input_ids=self.drafter.input_ids[:num_tokens],
+                        positions=self.drafter.positions[:num_tokens],
+                        hidden_states=self.drafter.hidden_states[:num_tokens],
+                    )
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu_model_runner.py:1506]: Can we move this to inside `eagle.py`?

Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:1506]: I mean a new method in the eagle proposer.

Comment (luyuzhe111) [reply vllm/v1/worker/gpu_model_runner.py]: @WoosukKwon done!

File: vllm/v1/spec_decode/eagle.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py:43]: Got it. Thanks for the clarification!

**Review by WoosukKwon — COMMENTED** [review]
@luyuzhe111 Thanks for addressing the comments! One last thing: Can you please add tests for this?

**Comment by luyuzhe111** [comment]
Hi @ekagra-ranjan, thanks for this [PR](https://github.com/vllm-project/vllm/pull/17323)! but wondering if you could create a new PR that makes sure the draft model takes in `vllm_config` instead of `model_config`. This is a requirement for compatibility with `torch.compile` decorator. thanks! cc @WoosukKwon

**Comment by luyuzhe111** [comment]
> One last thing: Can you please add tests for this?

@WoosukKwon I feel like acceptance length tests are probably the most meaningful tests for this PR. I tested the acceptance length by cherry picking commits from this [PR](https://github.com/vllm-project/vllm/pull/17010). You can see that the difference is less than 0.01. 

With `meta-llama/Llama-3.1-8B-Instruct`, `yuhuili/EAGLE-LLaMA3.1-Instruct-8B`

On MT Bench

When max number generated tokens = 256

| Number of Speculated Tokens | 1 | 2 | 3 | 4 | 5 |
|----------------------------|------|------|------|------|------|
| Eager | 1.71 | 2.09 | 2.30 | 2.38 | 2.43 |
| Compilation & CudaGraph | 1.70 | 2.10 | 2.29 | 2.38 | 2.43 |

I can do a follow-up PR adding acceptance length tests after https://github.com/vllm-project/vllm/pull/17010 is merged.

File: vllm/vllm_flash_attn/__init__.py
Comment (LucasWilkinson) [inline vllm/vllm_flash_attn/__init__.py:1]: Is adding this file a rebase error or is there a reason you are adding it?

**Review by LucasWilkinson — CHANGES_REQUESTED** [review]

File: vllm/v1/spec_decode/eagle.py [hunk 6]
```diff
@@ -26,10 +26,41 @@ def __init__(
         device: torch.device,
     ):
         self.vllm_config = vllm_config
+        self.method = self.vllm_config.speculative_config.method
         self.num_speculative_tokens = (
             vllm_config.speculative_config.num_speculative_tokens)
         self.max_model_len = vllm_config.model_config.max_model_len
         self.block_size = vllm_config.cache_config.block_size
+
+        self.dtype = vllm_config.model_config.dtype
```
Comment (ekagra-ranjan) [inline vllm/v1/spec_decode/eagle.py:35]: would this support target model in fp8 and eagle model in bf16? We could have situations when the target and draft model's have different dtype.

File: vllm/v1/spec_decode/eagle.py [hunk 7]
```diff
@@ -226,14 +290,20 @@ def load_model(self, target_model: nn.Module) -> None:
...
+            if self.vllm_config.speculative_config.method == "eagle":
+                self.model = draft_model_cls(
+                    vllm_config=self.vllm_config,
+                    start_layer_id=target_layer_num).to(target_device)
+            else:
+                assert self.vllm_config.speculative_config.method == "eagle3"
+                # TODO: make sure EAGLE3 models take in vllm_config as well
+                # for torch.compile compatibility.
+                self.model = draft_model_cls(
+                    model_config=draft_model_config,
```
Comment (ekagra-ranjan) [inline vllm/v1/spec_decode/eagle.py:302]: similar change of model_config -> vllm_config can be done for E3 model def and would resolve [this](https://github.com/vllm-project/vllm/pull/17211#issuecomment-2837460985)?

**Comment by ekagra-ranjan** [comment]
Thank you @luyuzhe111 for the PR!  Wondering if you have any benchmark for speedup in output token/s for with and w/o cuda graph and torch compile for llama 3.1 on MTBench TP1 BS1?

File: vllm/vllm_flash_attn/__init__.py
Comment (luyuzhe111) [reply vllm/vllm_flash_attn/__init__.py:1]: @LucasWilkinson sorry about the confusion. it was a rebase error. will revert.

File: vllm/v1/spec_decode/eagle.py (hunk 6 above)
Comment (luyuzhe111) [reply vllm/v1/spec_decode/eagle.py:35]: actually no since I forced the dtype to be that of the target model. for your case, i think it would be easier to just quantize the bf16 checkpoints?

File: vllm/v1/spec_decode/eagle.py (hunk 7 above)
Comment (luyuzhe111) [reply vllm/v1/spec_decode/eagle.py]: Yes indeed but I was worried the changes in Eagle3 model might get people confused. But given the discussion is already here, I have made similar changes to Eagle3 model. Thanks for the suggestion!

**Comment by luyuzhe111** [comment]
Hey @ekagra-ranjan Re: 

1. I remember you had some numbers for benchmarking on mt bench. can you share the setup so that I can run the benchmarking again for torch.compile + cuda graph? thanks!
2. for the forward pass I had to look at the profiler, so no script here. 
3. right I meant those operations that prepare inputs for the EAGLE model.

**Review by WoosukKwon — APPROVED** [review]
LGTM. @youkaichao Any final comments?

File: vllm/v1/spec_decode/eagle.py (hunk 6 above)
Comment (Swipe4057) [reply vllm/v1/spec_decode/eagle.py:35]: If my checkpoint is a model with static fp8, should my draft model also be fp8?

**Comment by ekagra-ranjan** [comment]
@luyuzhe111 - [here](https://github.com/vllm-project/vllm/pull/16937#issuecomment-2828889840) is an example setup for benchmark. Looking fwd to the results!

File: vllm/v1/spec_decode/eagle.py (hunk 6 above)
Comment (ekagra-ranjan) [reply vllm/v1/spec_decode/eagle.py:35]: With the current change it would need to be the same but if we could use the dtype from the draft model config then we add the flexibility of having target and draft model with different quantization methods. 

I think it would be more convinient to let them be separate which would allow plug and play different quantized format for target and draft. Wdyt @luyuzhe111 ?

**Review by LucasWilkinson — APPROVED** [review]

File: vllm/v1/spec_decode/eagle.py (hunk 6 above)
Comment (WoosukKwon) [reply vllm/v1/spec_decode/eagle.py:35]: @ekagra-ranjan I think supporting different quantization configs would be pretty complicated because the eagle head use the vocab embeddings (and sometimes even lm head) with the original model?

Anyways, I think we should consider it in a followup PR rather than this PR.

**Comment by luyuzhe111** [comment]
@ekagra-ranjan I got the following results:

Target model: `meta-llama/Llama-3.1-8B-Instruct`
EAGLE model: `yuhuili/EAGLE-LLaMA3.1-Instruct-8B`
Hardware: A100 (40GB)
Script: `VLLM_USE_V1=1 python examples/offline_inference/eagle.py —dataset="./data/mt_bench/question.jsonl" —num_spec_tokens x —max_num_seqs 1 —num_prompts 80`

Regular Decoding OTPS: 72
Screenshot 2025-04-29 at 3 21 29 PM

it looks like a further 10% speedup.

File: vllm/v1/spec_decode/eagle.py (hunk 6 above)
Comment (ekagra-ranjan) [reply vllm/v1/spec_decode/eagle.py:35]: I thought most quant recipies do not quantize lm_head or embed because it hurts the quality and it is better to have the logits on higher precision so that logprob are better.

**Comment by zou3519** [comment]
> One notable bug related to torch.compile is this [line](https://github.com/vllm-project/vllm/pull/17211/files#diff-a4809a837fbf535a8f0999b11087a53ec1c53948b50c0a1fe64396bc86de9461R153). Essentially, the default data type for input ids is int32 and the EAGLE model was also compiled with this data type. However, tensor.argmax() returns int64 by default. Feeding int64 input ids to the compiled model will completely mess things up and lead to gibberish draft tokens. Currently the compiled model does not even give warnings when the input data type mismatches. Wonder if we can prevent similar bugs in the future by some more checks.

These checks lead to performance degradation so vLLM decided to drop all of them. Some of the lower compilation_levels (e.g. 1) should do the checking but those also have slightly different behavior for other things

File: vllm/compilation/backends.py [hunk 8]
```diff
@@ -408,8 +412,13 @@ def __call__(self, graph: fx.GraphModule, example_inputs) -> Callable:
...
             self.compilation_config.cache_dir = cache_dir
 
-        cache_dir = self.compilation_config.cache_dir
+        if compilation_counter.num_graphs_seen > 0:
+            cache_dir = self.compilation_config.cache_dir + \
+                f'-{compilation_counter.num_graphs_seen}'
+        else:
+            cache_dir = self.compilation_config.cache_dir
         os.makedirs(cache_dir, exist_ok=True)
+        self.compilation_config.cache_dir = cache_dir
```
Comment (zou3519) [inline vllm/compilation/backends.py:421]: This is suspicious. Why do you need a different cache directory for each graph?

Comment (luyuzhe111) [reply vllm/compilation/backends.py:421]: @zou3519 thanks for reviewing! if there isn't a separate cache directory, the compiled code for the draft model (EAGLE) will not be saved at all. for models without EAGLE, my understanding is that the backend is invoked only once so this should not impact other models.

Comment (zou3519) [reply vllm/compilation/backends.py:421]: @luyuzhe111 thanks for the response and clarifying that. Woosuk also filled me in on some more details offline. I understand why we need a separate cache directory.

Which of the "original model" and the "eagle head" get compiled first? (I'm trying to figure out if the first cache dir is for the original model or for the eagle head)

Comment (luyuzhe111) [reply vllm/compilation/backends.py:421]: @zou3519 original model should be compiled first! also if you wanna double check, the transformed code of EAGLE in the cache directory has a slightly different signature with `hidden_states` as an additional arg. if there is a more elegant solution, that would be great! I think my approach is a bit hacky indeed : )))

Comment (ekagra-ranjan) [reply vllm/compilation/backends.py:421]: @zou3519 - Could you pls share the additional details which you got to learn from Woosuk? I still could not follow the need for a separate cache. Thank you!

Comment (zou3519) [reply vllm/compilation/backends.py:421]: Thanks for the discussion! I added some comments and an assertion into https://github.com/vllm-project/vllm/pull/17662 , please take a look.

I think in the future we'll want a better way to handle multiple compiled regions in a vLLM model, but that will take some re-designing

Comment (zou3519) [reply vllm/compilation/backends.py:421]: @luyuzhe111 the asserts in https://github.com/vllm-project/vllm/pull/17662 triggered, which means that this PR does affect non-eagle models

Comment (luyuzhe111) [reply vllm/compilation/backends.py:421]: @zou3519 Thanks for the catch! I guess a simple fix would be just to create a separate cache directory only for EAGLE, via looking at the vllm speculative config, for example?

Comment (zou3519) [reply vllm/compilation/backends.py:421]: Yeah that would work

File: vllm/compilation/backends.py [hunk 9]
```diff
@@ -408,8 +412,13 @@ def __call__(self, graph: fx.GraphModule, example_inputs) -> Callable:
             )
             self.compilation_config.cache_dir = cache_dir
 
-        cache_dir = self.compilation_config.cache_dir
+        if compilation_counter.num_graphs_seen > 0:
+            cache_dir = self.compilation_config.cache_dir + \
+                f'-{compilation_counter.num_graphs_seen}'
```
Comment (youkaichao) [inline vllm/compilation/backends.py:417]: I think we should be able to get the component's `prefix` to use as the cache directory, it could be more meaningful.

Comment (zou3519) [reply vllm/compilation/backends.py:417]: I'll make that change, @luyuzhe111 and I were discussing something similar above

Comment (zou3519) [reply vllm/compilation/backends.py:417]: https://github.com/vllm-project/vllm/pull/19027
